confidential-identity/src/claim_proofs.rs

```diff
@@ -93,167 +94,227 @@ impl ScopeClaimData {
-///
-/// # Inputs
-/// * `cdd_claim` is the CDD claim from which to generate the CDD_ID
-///
-/// # Output
-/// The Pedersen commitment result.
-pub fn compute_cdd_id(cdd_claim: &CddClaimData) -> CddId {
-    cryptography_core::cdd_claim::compute_cdd_id(cdd_claim)
-}
+const SIGNATURE_MESSAGE: &str = "SCOPE_ID is Wellformed";
 
-/// Compute the SCOPE_ID \
-/// SCOPE_ID = PedersenCommitment(SCOPE_DID, INVESTOR_UNIQUE_ID, [SCOPE_DID | INVESTOR_UNIQUE_ID])
-///
-/// # Inputs
-/// * `scope_claim` is the scope claim from which to generate the SCOPE_ID
-///
-/// # Output
-/// The Pedersen commitment result.
-pub fn compute_scope_id(scope_claim: &ScopeClaimData) -> RistrettoPoint {
-    generate_pedersen_commit(scope_claim.scope_did, scope_claim.investor_unique_id)
-}
+// -------------------------------------------------------------------------------------------
+// -                                Trait Implementations                                    -
+// -------------------------------------------------------------------------------------------
 
-pub fn build_scope_claim_proof_data(
-    cdd_claim: &CddClaimData,
-    scope_claim: &ScopeClaimData,
-) -> ScopeClaimProofData {
-    ScopeClaimProofData {
-        scope_did: scope_claim.scope_did,
-        investor_unique_id: cdd_claim.investor_unique_id,
-        investor_did: cdd_claim.investor_did,
+/// Implements the APIs of the CDD provider.
+pub struct Provider;
+
+impl ProviderTrait for Provider {
+    fn create_cdd_id(cdd_claim: &CddClaimData) -> CddId {
+        cryptography_core::cdd_claim::compute_cdd_id(cdd_claim)
     }
 }
 
-impl From<ScopeClaimProofData> for ProofKeyPair {
-    /// Create a key pair object for the investor from a claim data.
-    ///
-    /// # Input:
-    /// `d`: the data required to prove that a SCOPE_ID matches a CDD_ID.
-    fn from(d: ScopeClaimProofData) -> Self {
-        use sha3::{digest::FixedOutput, Digest, Sha3_256, Sha3_512};
-        // Investor's secret key is:
-        // Hash([INVESTOR_DID | INVESTOR_UNIQUE_ID]) - Hash([SCOPE_DID | INVESTOR_UNIQUE_ID])
-        let first_term = Scalar::from_hash(
-            Sha3_512::default()
-                .chain(d.investor_did.as_bytes())
-                .chain(d.investor_unique_id.as_bytes()),
+/// Implements the APIs of the Investor.
+pub struct Investor;
+
+impl InvestorTrait for Investor {
+    fn create_scope_claim_proof<R: RngCore + CryptoRng>(
+        cdd_claim: &CddClaimData,
+        scope_claim: &ScopeClaimData,
+        cdd_id: &CddId,
+        rng: &mut R,
+    ) -> ScopeClaimProof {
+        let scope_did_hash = slice_to_ristretto_point(scope_claim.scope_did.as_bytes());
+        let scope_id = scope_claim.investor_unique_id * scope_did_hash;
+
+        let (proof_scope_id_wellfromed, public) = sign(
+            &scope_claim.investor_unique_id,
+            &scope_did_hash,
+            SIGNATURE_MESSAGE.to_string(),
```

Comment:
In the write-up we talk about signing either a nonce or hash(scope_id) - are there any downsides to using a fixed message instead?

Reply:
Yeah the signature needs a nonce. I am in the process of implementing the signature as is done in https://docs.rs/ed25519-dalek/1.0.1/src/ed25519_dalek/secret.rs.html#389 which should have all the proper zeroizing and nonces.

Reply:
Pushed the updated version which zeroizes the secret key and uses secret_key as the seed for generating the nonce and other random values as done in `ed25519-dalek` (the schnorkel library does the same).

A note on the fixed message: while nonce IS important to the security of the protocol, the message is not important to the security of the protocol (@ajivanyan to confirm).

 If there is a business value in letting the users choose their message, we can change it and accept it as input.

Reply:
The message is not important to the security of the protocol. Using a fixed message for all signatures does not add any communication complexity to the proofs as well and I think no need to switch to custom messages.

Reply:
Thanks - makes sense to me. The verification side will be easier if there is a fixed message (or at least one which is deterministic from other known inputs so users don't need to enter anything).